[BUG] Nightly tests/test_umap.py::test_umap_model_persistence[False] FAILED

Details:
#spark-rapids-ml_nightly/590/

13:42:54  =================================== FAILURES ===================================
13:42:54  ______________________ test_umap_model_persistence[False] ______________________
13:42:54  [gw0] linux -- Python 3.10.16 /root/miniconda3/bin/python3.10
13:42:54  
13:42:54  sparse_fit = False, gpu_number = 1
13:42:54  tmp_path = '/tmp/spark_rapids_ml_tests_a72p1t1t'
13:42:54  
13:42:54      @pytest.mark.parametrize("sparse_fit", [True, False])
13:42:54      def test_umap_model_persistence(
13:42:54          sparse_fit: bool, gpu_number: int, tmp_path: str
13:42:54      ) -> None:
13:42:54          import pyspark
13:42:54          from packaging import version
13:42:54      
13:42:54          with CleanSparkSession() as spark:
13:42:54      
13:42:54              n_rows = 5000
13:42:54              n_cols = 200
13:42:54      
13:42:54              if sparse_fit:
13:42:54                  if version.parse(pyspark.__version__) < version.parse("3.4.0"):
13:42:54                      import logging
13:42:54      
13:42:54                      err_msg = "pyspark < 3.4 is detected. Cannot import pyspark `unwrap_udt` function for SparseVector. "
13:42:54                      "The test case will be skipped. Please install pyspark>=3.4."
13:42:54                      logging.info(err_msg)
13:42:54                      return
13:42:54      
13:42:54                  data, input_raw_data = _load_sparse_binary_data(n_rows, n_cols, 30)
13:42:54                  df = spark.createDataFrame(data, ["features"])
13:42:54              else:
13:42:54                  X, _ = make_blobs(
13:42:54                      n_rows,
13:42:54                      n_cols,
13:42:54                      centers=5,
13:42:54                      cluster_std=0.1,
13:42:54                      dtype=cp.float32,
13:42:54                      random_state=10,
13:42:54                  )
13:42:54                  pyspark_type = "float"
13:42:54                  feature_cols = [f"c{i}" for i in range(X.shape[1])]
13:42:54                  schema = [f"{c} {pyspark_type}" for c in feature_cols]
13:42:54                  df = spark.createDataFrame(X.tolist(), ",".join(schema))
13:42:54                  df = df.withColumn("features", array(*feature_cols)).drop(*feature_cols)
13:42:54                  input_raw_data = X.get()
13:42:54      
13:42:54              umap = UMAP(num_workers=gpu_number).setFeaturesCol("features")
13:42:54      
13:42:54              umap_model = umap.fit(df)
13:42:54              _assert_umap_model(umap_model, input_raw_data)
13:42:54      
13:42:54              # Model persistence
13:42:54              path = tmp_path + "/umap_tests"
13:42:54              model_path = f"{path}/umap_model"
13:42:54  >           umap_model.write().overwrite().save(model_path)
13:42:54  
13:42:54  tests/test_umap.py:461: 
13:42:54  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
13:42:54  /root/miniconda3/lib/python3.10/site-packages/pyspark/ml/util.py:140: in save
13:42:54      self.saveImpl(path)
13:42:54  src/spark_rapids_ml/umap.py:1492: in saveImpl
13:42:54      write_dense_array(array, df_path)
13:42:54  src/spark_rapids_ml/umap.py:1462: in write_dense_array
13:42:54      data_df = spark.createDataFrame(array)
13:42:54  /root/miniconda3/lib/python3.10/site-packages/pyspark/sql/session.py:894: in createDataFrame
13:42:54      return self._create_dataframe(
13:42:54  /root/miniconda3/lib/python3.10/site-packages/pyspark/sql/session.py:936: in _create_dataframe
13:42:54      rdd, struct = self._createFromLocal(map(prepare, data), schema)
13:42:54  /root/miniconda3/lib/python3.10/site-packages/pyspark/sql/session.py:631: in _createFromLocal
13:42:54      struct = self._inferSchemaFromList(data, names=schema)
13:42:54  /root/miniconda3/lib/python3.10/site-packages/pyspark/sql/session.py:517: in _inferSchemaFromList
13:42:54      schema = reduce(
13:42:54  /root/miniconda3/lib/python3.10/site-packages/pyspark/sql/session.py:519: in <genexpr>
13:42:54      (_infer_schema(row, names, infer_dict_as_struct, prefer_timestamp_ntz) for row in data),
13:42:54  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
13:42:54  
13:42:54  row = array([ -1.8275499, -10.793575 ], dtype=float32), names = None
13:42:54  infer_dict_as_struct = False, prefer_timestamp_ntz = False
13:42:54  
13:42:54      def _infer_schema(
13:42:54          row: Any,
13:42:54          names: Optional[List[str]] = None,
13:42:54          infer_dict_as_struct: bool = False,
13:42:54          prefer_timestamp_ntz: bool = False,
13:42:54      ) -> StructType:
13:42:54          """Infer the schema from dict/namedtuple/object"""
13:42:54          items: Iterable[Tuple[str, Any]]
13:42:54          if isinstance(row, dict):
13:42:54              items = sorted(row.items())
13:42:54      
13:42:54          elif isinstance(row, (tuple, list)):
13:42:54              if hasattr(row, "__fields__"):  # Row
13:42:54                  items = zip(row.__fields__, tuple(row))  # type: ignore[union-attr]
13:42:54              elif hasattr(row, "_fields"):  # namedtuple
13:42:54                  items = zip(row._fields, tuple(row))  # type: ignore[union-attr]
13:42:54              else:
13:42:54                  if names is None:
13:42:54                      names = ["_%d" % i for i in range(1, len(row) + 1)]
13:42:54                  elif len(names) < len(row):
13:42:54                      names.extend("_%d" % i for i in range(len(names) + 1, len(row) + 1))
13:42:54                  items = zip(names, row)
13:42:54      
13:42:54          elif hasattr(row, "__dict__"):  # object
13:42:54              items = sorted(row.__dict__.items())
13:42:54      
13:42:54          else:
13:42:54  >           raise TypeError("Can not infer schema for type: %s" % type(row))
13:42:54  E           TypeError: Can not infer schema for type: <class 'numpy.ndarray'>
13:42:54  
13:42:54  /root/miniconda3/lib/python3.10/site-packages/pyspark/sql/types.py:1302: TypeError

--- a/python/src/spark_rapids_ml/umap.py
+++ b/python/src/spark_rapids_ml/umap.py
@@ -48,12 +48,12 @@ from pyspark.ml.param.shared import (
 from pyspark.ml.util import DefaultParamsReader, DefaultParamsWriter, MLReader, MLWriter
 from pyspark.sql import Column, DataFrame
 from pyspark.sql.dataframe import DataFrame
-from pyspark.sql.functions import monotonically_increasing_id
 from pyspark.sql.types import (
     ArrayType,
     DoubleType,
     FloatType,
     IntegerType,
+    LongType,
     Row,
     StructField,
     StructType,
@@ -1436,20 +1436,49 @@ class _CumlModelWriterParquet(_CumlModelWriter):
     """
 
     def saveImpl(self, path: str) -> None:
+        from packaging import version
 
         spark = _get_spark_session()
+        require_schemas = version.parse(pyspark.__version__) < version.parse("3.4.0")
 
         def write_sparse_array(array: scipy.sparse.spmatrix, df_dir: str) -> None:
-            indptr_df = spark.createDataFrame(array.indptr, schema=["indptr"])
-            indices_data_df = spark.createDataFrame(
-                pd.DataFrame(
-                    {
-                        "indices": array.indices,
-                        "data": array.data,
-                        "row_id": range(len(array.indices)),
-                    }
+            # Spark < 3.4.0 cannot infer schemas from numpy arrays in createDataFrame.
+            if require_schemas:
+                indptr_schema = StructType(
+                    [StructField("indptr", IntegerType(), False)]
                 )
-            )
+                indptr_df = spark.createDataFrame(
+                    pd.DataFrame(array.indptr), schema=indptr_schema
+                )
+
+                indices_data_schema = StructType(
+                    [
+                        StructField("indices", IntegerType(), False),
+                        StructField("data", FloatType(), False),
+                        StructField("row_id", LongType(), False),
+                    ]
+                )
+                indices_data_df = spark.createDataFrame(
+                    pd.DataFrame(
+                        {
+                            "indices": array.indices,
+                            "data": array.data,
+                            "row_id": range(len(array.indices)),
+                        }
+                    ),
+                    schema=indices_data_schema,
+                )
+            else:
+                indptr_df = spark.createDataFrame(array.indptr, schema="indptr: int")
+                indices_data_df = spark.createDataFrame(
+                    pd.DataFrame(
+                        {
+                            "indices": array.indices,
+                            "data": array.data,
+                            "row_id": range(len(array.indices)),
+                        }
+                    )
+                )
 
             indptr_df.write.parquet(
                 os.path.join(df_dir, "indptr.parquet"), mode="overwrite"
@@ -1459,7 +1488,17 @@ class _CumlModelWriterParquet(_CumlModelWriter):
             )
 
         def write_dense_array(array: np.ndarray, df_path: str) -> None:
-            data_df = spark.createDataFrame(array)
+            # Spark < 3.4.0 cannot infer schemas from numpy arrays in createDataFrame.
+            if require_schemas:
+                schema = StructType(
+                    [
+                        StructField(f"_{i}", FloatType(), False)
+                        for i in range(1, array.shape[1] + 1)
+                    ]
+                )
+                data_df = spark.createDataFrame(pd.DataFrame(array), schema=schema)
+            else:
+                data_df = spark.createDataFrame(array)
             data_df.write.parquet(df_path, mode="overwrite")
 
         DefaultParamsWriter.saveMetadata(

--- a/python/src/spark_rapids_ml/umap.py
+++ b/python/src/spark_rapids_ml/umap.py
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2024, NVIDIA CORPORATION.
+# Copyright (c) 2025, NVIDIA CORPORATION.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.

--- a/python/src/spark_rapids_ml/umap.py
+++ b/python/src/spark_rapids_ml/umap.py
@@ -1436,49 +1436,32 @@ class _CumlModelWriterParquet(_CumlModelWriter):
     """
 
     def saveImpl(self, path: str) -> None:
-        from packaging import version
 
         spark = _get_spark_session()
-        require_schemas = version.parse(pyspark.__version__) < version.parse("3.4.0")
 
         def write_sparse_array(array: scipy.sparse.spmatrix, df_dir: str) -> None:
-            # Spark < 3.4.0 cannot infer schemas from numpy arrays in createDataFrame.
-            if require_schemas:
-                indptr_schema = StructType(
-                    [StructField("indptr", IntegerType(), False)]
-                )
-                indptr_df = spark.createDataFrame(
-                    pd.DataFrame(array.indptr), schema=indptr_schema
-                )
+            indptr_schema = StructType([StructField("indptr", IntegerType(), False)])
+            indptr_df = spark.createDataFrame(
+                pd.DataFrame(array.indptr), schema=indptr_schema
+            )
 
-                indices_data_schema = StructType(
-                    [
-                        StructField("indices", IntegerType(), False),
-                        StructField("data", FloatType(), False),
-                        StructField("row_id", LongType(), False),
-                    ]
-                )
-                indices_data_df = spark.createDataFrame(
-                    pd.DataFrame(
-                        {
-                            "indices": array.indices,
-                            "data": array.data,
-                            "row_id": range(len(array.indices)),
-                        }
-                    ),
-                    schema=indices_data_schema,
-                )
-            else:
-                indptr_df = spark.createDataFrame(array.indptr, schema="indptr: int")
-                indices_data_df = spark.createDataFrame(
-                    pd.DataFrame(
-                        {
-                            "indices": array.indices,
-                            "data": array.data,
-                            "row_id": range(len(array.indices)),
-                        }
-                    )
-                )
+            indices_data_schema = StructType(
+                [
+                    StructField("indices", IntegerType(), False),
+                    StructField("data", FloatType(), False),
+                    StructField("row_id", LongType(), False),
+                ]
+            )
+            indices_data_df = spark.createDataFrame(
+                pd.DataFrame(
+                    {
+                        "indices": array.indices,
+                        "data": array.data,
+                        "row_id": range(len(array.indices)),
+                    }
+                ),
+                schema=indices_data_schema,
+            )
 
             indptr_df.write.parquet(
                 os.path.join(df_dir, "indptr.parquet"), mode="overwrite"
@@ -1488,17 +1471,13 @@ class _CumlModelWriterParquet(_CumlModelWriter):
             )
 
         def write_dense_array(array: np.ndarray, df_path: str) -> None:
-            # Spark < 3.4.0 cannot infer schemas from numpy arrays in createDataFrame.
-            if require_schemas:
-                schema = StructType(
-                    [
-                        StructField(f"_{i}", FloatType(), False)
-                        for i in range(1, array.shape[1] + 1)
-                    ]
-                )
-                data_df = spark.createDataFrame(pd.DataFrame(array), schema=schema)
-            else:
-                data_df = spark.createDataFrame(array)
+            schema = StructType(
+                [
+                    StructField(f"_{i}", FloatType(), False)
+                    for i in range(1, array.shape[1] + 1)
+                ]
+            )
+            data_df = spark.createDataFrame(pd.DataFrame(array), schema=schema)
             data_df.write.parquet(df_path, mode="overwrite")
 
         DefaultParamsWriter.saveMetadata(

